File: src/main/java/com/hp/jipp/encoding/KeyValueType.java
package com.hp.jipp.encoding;

import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class KeyValueType extends AttributeType<Map<String, String>> {

    /** An encoder for Key/Value types */
    public static final Attribute.SimpleEncoder<Map<String, String>> ENCODER =
            new Attribute.SimpleEncoder<Map<String, String>>("KeyValue") {
                static final String ELEMENT_SEPARATOR = ";";
            static final String PART_SEPARATOR = "=";

            @Override
            public Map<String, String> readValue(DataInputStream in, Tag valueTag) throws IOException {
                String input = StringType.ENCODER.readValue(in, valueTag);
                return decode(input);
            }

            @Override
            public void writeValue(DataOutputStream out, Map<String, String> value) throws IOException {
                StringType.ENCODER.writeValue(out, encode(value));
            }

            @Override
            public boolean valid(Tag valueTag) {
                return valueTag == Tag.OctetString;
            }

            private Map<String, String> decode(String input) {
                Map<String, String> map = new LinkedHashMap<>();
                for (String element : Splitter.on(ELEMENT_SEPARATOR).split(input)) {
                    List<String> parts = ImmutableList.copyOf(Splitter.on(PART_SEPARATOR).split(element));
                    if (parts.size() == 2 && !parts.get(0).isEmpty() && !parts.get(1).isEmpty()) {
                        map.put(parts.get(0), parts.get(1));
                    }
                }
                return ImmutableMap.copyOf(map);
            }

            private String encode(Map<String, String> input) {
                StringBuilder out = new StringBuilder();
                for (Map.Entry<String, String> entry : input.entrySet()) {
                    out.append(entry.getKey());
                    out.append(PART_SEPARATOR);
                    out.append(entry.getValue());
                    out.append(ELEMENT_SEPARATOR);
                }
                return out.toString();
            }
        };

    public KeyValueType(String name) {
        super(ENCODER, Tag.OctetString, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/KeyValueType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import java.io.DataInputStream
import java.io.DataOutputStream

/** Attribute type for Key/Value pairs */
class KeyValueType(override val name: String) : AttributeType<Map<String, String>>(Encoder, Tag.octetString) {
    companion object Encoder : SimpleEncoder<Map<String, String>>("KeyValue") {
        private const val ELEMENT_SEPARATOR = ";"
        private const val PART_SEPARATOR = "="

        override fun writeValue(out: DataOutputStream, value: Map<String, String>) =
                StringType.Encoder.writeValue(out, encode(value))

        private fun encode(input: Map<String, String>): String {
            val out = StringBuilder()
            input.forEach {
                out.append(it.key)
                out.append(PART_SEPARATOR)
                out.append(it.value)
                out.append(ELEMENT_SEPARATOR)
            }
            return out.toString()
        }

        override fun valid(valueTag: Tag) = valueTag == Tag.octetString

        override fun readValue(input: DataInputStream, valueTag: Tag): Map<String, String> =
                decode(StringType.Encoder.readValue(input, valueTag))

        private fun decode(input: String): Map<String, String> =
                input.split(ELEMENT_SEPARATOR)
                        .map { it.split(PART_SEPARATOR) }
                        .filter { it.size == 2 && it[0].isNotEmpty() && it[1].isNotEmpty() }
                        .map { it[0] to it[1] }
                        .toMap()
    }

    // Include these to allow java to see the correct types instead of Map<String, ? extends String>

    override fun of(value: Map<String, String>, vararg values: Map<String, String>): Attribute<Map<String, String>> =
            super.invoke(listOf(value) + values.toList())

    override fun of(values: List<Map<String, String>>): Attribute<Map<String, String>> =
            super.invoke(values)
}
